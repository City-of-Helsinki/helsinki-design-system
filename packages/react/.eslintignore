**/*.d.ts
webpack.config.js
.templates
lib/
**/*.stories.tsx
