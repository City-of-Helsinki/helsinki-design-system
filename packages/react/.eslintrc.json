{
  "parser": "@typescript-eslint/parser",
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:react/recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:import/errors",
    "plugin:import/warnings",
    "plugin:import/typescript",
    "plugin:jest/recommended",
    "airbnb",
    "prettier/@typescript-eslint",
    "plugin:prettier/recommended" // this should always be the last item on this array
  ],
  "plugins": ["react", "@typescript-eslint", "prettier", "jest", "import"],
  "parserOptions": {
    "ecmaVersion": 2018,
    "ecmaFeatures": {
      "jsx": true
    }
  },
  "ignorePatterns": ["*.module.css"],
  "rules": {
    "no-use-before-define": "off",
    "@typescript-eslint/no-use-before-define": ["error"],
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": ["**/*.test.ts", "**/*.test.tsx", "**/test*.ts*", "**/*.stories.tsx"]
      }
    ],
    "import/order": [
      "error",
      {
        "groups": ["builtin", "external", "internal", ["parent", "sibling", "index"]],
        "newlines-between": "always"
      }
    ],
    "react/jsx-filename-extension": "off",
    "react/jsx-one-expression-per-line": "off",
    "react/jsx-props-no-spreading": "off",
    "react/jsx-curly-newline": "off",
    "@typescript-eslint/explicit-function-return-type": "off",
    "react/prop-types": "off",
    "import/extensions": "off",
    "import/prefer-default-export": "off",
    "jsx-a11y/label-has-associated-control": [
      2,
      {
        "components": ["Label"],
        "required": {
          "some": ["nesting", "id"]
        },
        "allowChildren": false
      }
    ],
    "@typescript-eslint/explicit-member-accessibility": "off",
    "react/state-in-constructor": [2, "never"],
    "no-unused-expressions": [2, {
      "allowTernary": true
    }],
    "no-unused-vars": "off",
    "@typescript-eslint/no-unused-vars": ["error"]
  },
  "settings": {
    "react": {
      "version": "detect"
    },
    "import/resolver": {
      "node": {
        "extensions": [".js", ".jsx", ".ts", ".tsx"],
        "paths": ["src"]
      }
    }
  },
  "env": {
    "browser": true,
    "jest": true
  },
  "globals": {
    "JSX": true
  }
}
