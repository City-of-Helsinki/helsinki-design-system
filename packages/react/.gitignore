node_modules/
storybook-static/
lib/
_bundles/
*.css.d.ts
.loki/current
.loki/difference

*storybook.log
