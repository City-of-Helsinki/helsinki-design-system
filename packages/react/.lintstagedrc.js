module.exports = {
  'src/**/*.{ts,tsx}': 'eslint',
}
