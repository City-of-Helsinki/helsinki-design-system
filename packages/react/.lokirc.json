{
  "chromeRetries": 2,
  "configurations": {
    "chrome.laptop": {
      "target": "chrome.docker",
      "width": 1366,
      "height": 768,
      "mobile": false
    },
    "chrome.iphone7": {
      "target": "chrome.docker",
      "preset": "iPhone 7"
    }
  },
  "skipStories": "Icons/.*|Button/*|Tag/*|Hero/*|Koros/*|Checkbox/*",
  "pixelmatch": {
    "threshold": 0.2,
    "alpha": 1,
    "diffColor": [255, 0, 255]
  }
}
