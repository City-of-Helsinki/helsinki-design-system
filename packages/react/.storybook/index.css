@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/565d73a693abe0776c801607ac28f0bf.woff') format('woff');
  font-display: swap;
  font-weight: 400;
  font-style: normal;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/5bb29e3b7b1d3ef30121229bbe67c3e1.woff') format('woff');
  font-display: swap;
  font-weight: 400;
  font-style: italic;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/7c46f288e8133b87e6b12b45dac71865.woff') format('woff');
  font-display: swap;
  font-weight: 500;
  font-style: normal;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/e62dc97e83a385e4d8cdc939cf1e4213.woff') format('woff');
  font-display: swap;
  font-weight: 500;
  font-style: italic;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/533af26cf28d7660f24c2884d3c27eac.woff') format('woff');
  font-display: swap;
  font-weight: 700;
  font-style: normal;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/20d494430c87e15e194932b729d48270.woff') format('woff');
  font-display: swap;
  font-weight: 700;
  font-style: italic;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/a50a1bd245ce63abcc0d1da80ff790d2.woff') format('woff');
  font-display: swap;
  font-weight: 900;
  font-style: normal;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/62a1781d8b396fbb025b0552cf6304d2.woff') format('woff');
  font-display: swap;
  font-weight: 900;
  font-style: italic;
  text-rendering: optimizeLegibility;
}

body {
  font-family: var(--font-default);
  font-size: var(--fontsize-body-m);
  color: var(--color-black-90);
  line-height: var(--lineheight-l);
}

.password-input--external-show-password-button,
.date-input--external-clear-value-button {
  display: grid;
  grid-template-columns: 400px 200px;
  gap: 20px;
  align-items: center;
}

@media only screen and (max-width: 768px) {
  .password-input--external-show-password-button,
  .date-input--external-clear-value-button {
    grid-template-columns: 300px;
  }
}

@media only screen and (max-width: 575.98px) {
  .table-custom-action {
    width: 100%;
  }
}

/* STEPPER */

.stepper-small {
  max-width: 300px;
}

@media only screen and (max-width: 575.98px) {
  .stepper-small {
    max-width: 100%;
  }
}

.stepper-margin {
  margin-left: -35px;
}

.stepper-heading {
  margin-left: 10px;
}

.stepper-form-validation {
  margin-left: -18px;
}

.stepper-card > div:first-child {
  margin-bottom: var(--spacing-m) !important;
}

#focused-element-after-cookie-consent-closed:focus {
  outline: 2px black solid;
}
