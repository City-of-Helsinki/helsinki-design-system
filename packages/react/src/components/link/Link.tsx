import React from 'react';

// import core base styles
import 'hds-core';
import styles from './Link.module.scss';
import { IconLinkExternal } from '../../icons';
import classNames from '../../utils/classNames';
import { getTextFromReactChildren } from '../../utils/getTextFromReactChildren';

export type LinkProps = Omit<
  React.ComponentPropsWithoutRef<'a'>,
  'target' | 'href' | 'onPointerEnterCapture' | 'onPointerLeaveCapture' | 'aria-label'
> & {
  /**
   * Link content
   */
  children: string;
  /**
   * Boolean indicating whether visited styles of the link are applied
   */
  disableVisitedStyles?: boolean;
  /**
   * Boolean indicating whether the link will lead user to external domain.
   */
  external?: boolean;
  /**
   * Hypertext Reference of the link.
   */
  href: string;
  /**
   * Boolean indicating whether the link will open in new tab or not.
   */
  openInNewTab?: boolean;
  /**
   * Aria label for opening link in a new tab
   */
  openInNewTabAriaLabel?: string;
  /**
   * Aria label for opening link in an external domain
   */
  openInExternalDomainAriaLabel?: string;
  /**
   * Size of the link
   */
  size?: 'S' | 'M' | 'L';
  /**
   * Additional styles
   */
  style?: React.CSSProperties;
};

type LinkToIconSizeMappingType = {
  L: 'l';
  M: 's';
  S: 'xs';
};

export const Link = ({
  children,
  className,
  disableVisitedStyles = false,
  external = false,
  href,
  openInNewTab = false,
  openInExternalDomainAriaLabel,
  openInNewTabAriaLabel,
  style = {},
  size = 'S',
  ...rest
}: LinkProps) => {
  const comboseAriaLabel = () => {
    let childrenText = getTextFromReactChildren(children);
    const newTabText = openInNewTab ? openInNewTabAriaLabel || 'Avautuu uudessa välilehdessä.' : '';
    const externalText = external ? openInExternalDomainAriaLabel || 'Avautuu uuteen domainiin.' : '';

    if (childrenText && childrenText.slice(-1) !== '.') {
      childrenText = `${childrenText}.`;
    }

    return [childrenText, newTabText, externalText].filter((text) => text).join(' ');
  };

  const mapLinkSizeToExternalIconSize: LinkToIconSizeMappingType = {
    L: 'l',
    M: 's',
    S: 'xs',
  };

  return (
    <a
      className={classNames(
        styles.link,
        styles[`link${size}`],
        disableVisitedStyles ? styles.disableVisitedStyles : '',
        className,
      )}
      href={href}
      style={style}
      {...(openInNewTab && { target: '_blank' })}
      {...((openInNewTab || external) && { 'aria-label': comboseAriaLabel() })}
      {...rest}
    >
      {children}
      {external && (
        <IconLinkExternal
          size={mapLinkSizeToExternalIconSize[size]}
          className={classNames(
            styles.icon,
            styles[size === 'L' ? 'verticalAlignBigIcon' : 'verticalAlignSmallOrMediumIcon'],
          )}
          aria-hidden
        />
      )}
    </a>
  );
};
