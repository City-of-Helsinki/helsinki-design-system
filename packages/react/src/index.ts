export * from './icons';
export * from './components';
export * from './ssr';
