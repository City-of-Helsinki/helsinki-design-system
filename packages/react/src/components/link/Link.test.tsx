import React from 'react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';

import { Link } from './Link';

describe('<Link /> spec', () => {
  it('renders the component', () => {
    const { asFragment } = render(<Link href="https://hds.hel.fi">Test link</Link>);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should not have basic accessibility issues', async () => {
    const { container } = render(<Link href="https://hds.hel.fi">Test link</Link>);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
