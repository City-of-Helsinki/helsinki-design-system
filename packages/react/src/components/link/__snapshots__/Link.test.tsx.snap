// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> spec renders the component 1`] = `
<DocumentFragment>
  <a
    class="link linkS"
    href="https://hds.hel.fi"
  >
    Test link
  </a>
</DocumentFragment>
`;
